docs/streams/developer-guide.html

```diff
@@ -267,21 +295,532 @@
+             // Java 7 example: cf. `map` for how to create `KeyValueMapper` instances
+         </pre>
+         </td>
+     </tr>
+       <tr>
+           <td><b>FlatMap (values only)</b>: <code>KStream->KStream </code></td>
+           <td>
+               Takes one record and produces zero, one, or more records, while retaining the key of the original record.
+               You can modify the record values and the value type.
+
+               <code>flatMapValues</code> is preferable to <code>flatMap</code> because it will not cause data re-partitioning. However,
+               it does not allow you to modify the key or key type like <code>flatMap</code> does.
+
+               <pre class="brush: java;">
+               // Split a sentence into words.
+               KStream&lt;byte[], String&gt; sentences = ...;
+               KStream&lt;byte[], String&gt; words = sentences.flatMapValues(value -> Arrays.asList(value.split("\\s+")));
+
+               // Java 7 example: cf. `mapValues` for how to create `ValueMapper` instances
+               </pre>
+           </td>
+       </tr>
+       <tr>
+           <td><b>Foreach</b>: <code>KStream->void or KTable->void </code></td>
+           <td>
+               Terminal operation. Performs a stateless action on each record.<p>
+
+               Note on processing guarantees: Any side effects of an action (such as writing to external systems)
+               are not trackable by Kafka, which means they will typically not benefit from Kafka's processing guarantees.
+
+               <pre class="brush: java;">
+               KStream&lt;String, Long&gt; stream = ...;
+
+               // Print the contents of the KStream to the local console.
+               // Java 8+ example, using lambda expressions
+               stream.foreach((key, value) -> System.out.println(key + " => " + value));
+
+               // Java 7 example
+               stream.foreach(
+                   new ForeachAction&lt;String, Long&gt;() {
+                       @Override
+                       public void apply(String key, Long value) {
+                         System.out.println(key + " => " + value);
+                       }
+               });
+               </pre>
+           </td>
+       </tr>
+       <tr>
+           <td><b>GroupByKey</b>: <code>KStream->KGroupedStream </code></td>
+           <td>
+               Groups the records by the existing key.<p>
+
+               Grouping is a prerequisite for aggregating a stream or a table and ensures that data is properly
+               partitioned ("keyed") for subsequent operations.<p>
+
+               <b>When to set explicit serdes</b>: Variants of <code>groupByKey</code> exist to override the configured default serdes of
+               your application, which you must do if the key and/or value types of the resulting <code>KGroupedStream</code> do
+               not match the configured default serdes.<p>
+
+               <b>Note:</b>
```

Comment:
Since we don't have a style/CSS for info boxes yet, I'd suggest to change this to:

```
<b>Note on Grouping vs. Windowing:</b>
```